controllers/kubemarkcluster.go

```diff
@@ -0,0 +1,65 @@
+package controllers
+
+import (
+	gocontext "context"
+	"strings"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type KubemarkCluster interface {
+	GenerateKubemarkClusterClient(kubemarkClusterSecretRef *corev1.ObjectReference, ownerNamespace string, context gocontext.Context) (client.Client, string, error)
+}
+
+// NewKubemarkCluster creates new KubemarkCluster instance
+func NewKubemarkCluster(client client.Client) KubemarkCluster {
+	return &kubemarkCluster{
+		Client: client,
+	}
+}
+
+type kubemarkCluster struct {
+	client.Client
+}
+
+// GenerateKubemarkClusterClient creates a client for kubemark cluster.
+func (w *kubemarkCluster) GenerateKubemarkClusterClient(kubemarkClusterSecretRef *corev1.ObjectReference, ownerNamespace string, context gocontext.Context) (client.Client, string, error) {
+	if kubemarkClusterSecretRef == nil {
+		return w.Client, ownerNamespace, nil
```

Comment:
should this func return an error if `w.Client` is `nil` as well?

Reply:
`w.Client` should never be `null` here since we initialize it in controller `main.go` with controller-runtime client ... but doesn't hurt to be defensive

Reply:
if we know that `w.Client` won't be `nil` then i'm fine to leave it as is, i just wasn't sure if that was possible or expected.

Reply:
👍